### DISCUSSION TITLE
Definitions...

### DISCUSSION THREAD

**Discussion by SteffenBrinckmann** [marda-alliance/metadata_extractors#14] [General]
After recapping the 2nd meeting, I noticed that we should better define the word "extractor". Do we mean:
- 1:1 translation of a possibly proprietary binary file into a "NeXus file": 
  - I use Nexus as an example of a fileformat that is semi-defined and can grow in its description
  - 1:1 such that we maintain units, etc from the binary file (if distance is in inch, we do not convert)
  - I don't think this is what the majority thinks
- Data analysis and production of certain measures, data, metadata from a file
  - In yesterdays discussion we identified peaks, spoke about branching or recipes of production
  - This analysis and production will depend on the scientific question
  - Is this what the majority thinks?

Terms like 'energy' are well defined. I think we should watch out that we understand the same thing for common words that are not as defined.

**Comment by PeterKraus** [comment]
My understanding of the original scope of the project was to provide a framework for "automated extraction of metadata" from non-FAIR filetypes. The idea was that you'd at least be able to query a database to answer what quantities (such as 'position', 'time', or 'velocity') are stored in your black-box-files, and ideally getting a way of getting at them. This at least lets you know what data you have and then formulate a way how to calculate 'velocity' from your 'position' and 'time'. This "calculation" step is in my view out of scope of this project.

This of course runs into the issue of "what constitutes data/metadata", which is a difficult discussion. Things like "filename" and "author" are clearly metadata, perhaps also matrix/dataset sizes and even date ranges or locations. Things like "units" are debatable. My suggestion was (at least for `v0` of our schemas) to table this discussion, but perhaps it's better to have it now: My two cents are to try and split "translation into FAIR format" from "postprocessing", but I understand it's not often practicable, especially with large datasets.

As "extractor" I currently understand a software that's able to read data from the supported filetypes; and understands what that data is on some level. We tried to avoid the usage of the word "parser", as that implies some level of semantic translation, as opposed to just understanding the topology of the files in their native nomenclature. Providing a framework for exchange of any of this understanding (i.e. a part of the "metadata") would be tremendously useful.

Finally, I have recently started thinking of this project as "metadata of extractors" (interoperability of tools) rather than "extractors of metadata" (interoperability of and ontology in filetypes). I don't think we can have one without the other.

**Comment by SteffenBrinckmann** [comment]
I think an example might make it clearer. So you think it would be:
``` bash
executable spectraIron.xyz -o peaks
# returns: {"x":[1,3,6], "y":[10,14,4]}
```
Or would it be:
``` bash
executable spectraIron.xyz -o neXus
# returns nexus file spectraIron.h5
```

**Reply by PeterKraus** [reply]
The second one, yes:
```
tractor spectraIron.xyz --output NeXus
  -> internally queries for filetype of spectraIron.xyz = spectrum-xyz
  -> finds extractors for the spectrum-xyz filetype = [specx, peakxyz]
      -> selects extractor that can convert to "neXus" = specx
  -> executes the extractor according to usage = specx spectraIron.xyz spectraIron.h5
# returns nexus file spectraIron.h5
```
For this we need the registry to provide a mapping of `input->output` filetypes and `extractors` that provide them. Note that currently, we're working on the inputs.


The first one, only if `"peaks"` are a known quantity:
```
tractor spectraIron.xyz --get peaks
  -> internally queries for filetype of spectraIron.xyz = spectrum-xyz
  -> finds extractors for the spectrum-xyz filetype = [specx, peakxyz]
      -> selects extractor that can extract "peaks" = peakxyz
  -> executes the extractor according to usage = peakxyz --filetype=spectrum-xyz ... spectraIron.xyz
     # returns a giant object of some kind that has a defined and understood schema
     # uses defined output schema of extractor to find "peaks" in that object
# returns: {"x":[1,3,6], "y":[10,14,4]}
# or returns: {} if peaks not in file
```
For this, we need the registry to provide a way to understand the query in the first place, and the `extractors` must be self-describing enough that `tractor` knows which `extractor` can extract `peaks` from which `filetype`.
